testonly/hammer/replay.go

```diff
@@ -0,0 +1,182 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hammer
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/google/trillian"
+	"google.golang.org/grpc"
+)
+
+type recordingInterceptor struct {
+	mu     sync.Mutex
+	outLog io.Writer
+}
+
+// NewRecordingInterceptor returns a grpc.UnaryClientInterceptor that logs outgoing
+// requests to file.
+func NewRecordingInterceptor(filename string) (grpc.UnaryClientInterceptor, error) {
+	o, err := os.Create(filename)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create log file: %v", err)
+	}
+	ri := recordingInterceptor{outLog: o}
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		return ri.invoke(ctx, method, req, reply, cc, invoker, opts...)
+	}, nil
+}
+
+func (ri *recordingInterceptor) invoke(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+	if msg, ok := req.(proto.Message); ok {
+		ri.dumpMessage(msg)
+	} else {
+		glog.Warningf("failed to convert request %T to proto.Message", req)
+	}
+	err := invoker(ctx, method, req, reply, cc, opts...)
+	if err == nil {
+		if msg, ok := reply.(proto.Message); ok {
+			ri.dumpMessage(msg)
+		} else {
+			glog.Warningf("failed to convert response %T to proto.Message", req)
+		}
+	}
+	return err
+}
+
+func (ri *recordingInterceptor) dumpMessage(in proto.Message) {
+	ri.mu.Lock()
+	defer ri.mu.Unlock()
+	if err := writeMessage(ri.outLog, in); err != nil {
+		glog.Error(err.Error())
+	}
+}
+
+func writeMessage(w io.Writer, in proto.Message) error {
+	a, err := ptypes.MarshalAny(in)
+	if err != nil {
+		return fmt.Errorf("failed to marshal %T %+v to any.Any: %v", in, in, err)
+	}
+	data, err := proto.Marshal(a)
+	if err != nil {
+		return fmt.Errorf("failed to marshal any.Any: %v", err)
+	}
+	// Encode as [4-byte big-endian length, message]
+	lenData := make([]byte, 4)
+	binary.BigEndian.PutUint32(lenData, uint32(len(data)))
+	w.Write(lenData)
+	w.Write(data)
+	return nil
+}
+
+func readMessage(r io.Reader) (*any.Any, error) {
+	// Decode from [4-byte big-endian length, message]
+	var lenData [4]byte
+	dlen, err := r.Read(lenData[:])
+	if err == io.EOF {
+		return nil, io.EOF
+	}
+	if dlen < 4 {
+		return nil, fmt.Errorf("corrupt data: expected 4-byte length, only found %d bytes", dlen)
+	}
+	l := binary.BigEndian.Uint32(lenData[:])
+	data := make([]byte, l)
+	n, err := r.Read(data)
```

Comment:
I think you could simplify the Read, buffer decode steps into a single read from input stream like so:

 ```go
	var pi float64
	b := []byte{0x18, 0x2d, 0x44, 0x54, 0xfb, 0x21, 0x09, 0x40}
	buf := bytes.NewReader(b)
	err := binary.Read(buf, binary.LittleEndian, &pi)
```

Reply:
I don't understand what you're suggesting.  The data is encoded as {4-byte-len, encoded-protobuf) chunks, so the length has to be read before the encoded-protobuf can be safely pulled in, and this function reads a single (len, proto) pair at a time which would allow streaming use.

Reply:
Ah, that makes more sense after the edit -- done.